projects/dataset_recorder/dataset_recorder.cc

```diff
@@ -0,0 +1,172 @@
+
+using namespace BoBRobotics;
+using namespace std::literals;
+using namespace units::angle;
+using namespace units::length;
+using namespace units::literals;
+using namespace units::time;
+
+int
+bobMain(int argc, char *argv[])
+{
+    // setting up
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    Robots::RCCarBot bot;
+    GPS::Gps gps;
+    gps.connect(path_linux);
+    BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+
+    const millisecond_t runTime = (argc > 1) ? second_t{ std::stod(argv[1]) } : 30_s;
+    LOGD << "running for " << runTime;
+
+    // check to see if we get valid coordinates by checking GPS quality
+    GPS::GPSData gpsData;
+    for (; numTrials > 0; numTrials--) {
+        try {
+            gpsData = gps.getGPSData();
+        } catch (GPS::GPSError &e) {
+            LOGW << " measuring failed, trying again in 1 second "
+                 << "[" << maxTrials - numTrials << "/" << maxTrials << "]";
+            std::this_thread::sleep_for(1s);
+            continue;
+        }
+
+        if (gpsData.gpsQuality != GPS::GPSQuality::INVALID) {
+            LOGI << " we have a valid measurement";
+            break;
+        }
+    }
+
+    if (numTrials == 0) {
+        LOGW << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program ";
+        return EXIT_FAILURE;
+    }
+
+    // Use the system clock to get date (might be wrong!)
+    time_t rawtime;
+    time(&rawtime);
+    std::tm currentTime = *localtime(&rawtime);
+
+    /*
+     * Use GPS to get time, because Jetsons don't remember the system time
+     * across boots.
+     */
+    currentTime.tm_hour = gpsData.time.hour;
+    currentTime.tm_min = gpsData.time.minute;
+    currentTime.tm_sec = gpsData.time.second;
+
+    // Make a new image database using current time to generate folder name
+    Navigation::ImageDatabase database{ currentTime };
+
+    // PixPro defaults to 1440x1440
+    auto cam = Video::getPanoramicCamera();
+    LOGI << "camera initialised " << cam->getOutputSize();
+
+    BackgroundExceptionCatcher catcher;
+
+    Stopwatch sw;
+    std::array<degree_t, 3> angles;
+
+    auto recorder = database.getRouteVideoRecorder(cam->getOutputSize(),
+                                                   cam->getFrameRate(),
+                                                   "Pitch [degrees]",
+                                                   "Roll [degrees]",
+                                                   "Speed",
+                                                   "Steering angle [degrees]",
+                                                   "UTM zone",
+                                                   "GPS quality",
+                                                   "Timestamp [ms]");
+
+    cv::Mat frame;
+    catcher.trapSignals();
+    sw.start();
+    for (auto timestamp = 0_ms; timestamp < runTime; timestamp = sw.elapsed()) {
```

Comment:
to my mind timesteps are integers - maybe rename this to ``time``

Reply:
Ok cool